### DISCUSSION TITLE
Duration storage

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#5955] [Design Proposals]
Now that I've got our `Duration` support working properly again, it has re-occurred to me that the way Hibernate stores `Duration`s is not very user-friendly: sure, Java represents all durations with nanosecond precision, but it's not very nice to open up the database and see a value with 11 zeros on it representing one measly hour. A nanosecond is not actually a very useful human-scale measure of time.

So to my mind, a much friendlier way to store a duration is as seconds, i.e. using `numeric(21,9)` instead of `numeric(21)`. 

Or, hell, even as _days_ would make sense, using something like `numeric(22,14)`.

How could we accommodate this in a way that doesn't break backward compatibility with our own data?

Well, I know I've resisted introducing new special-purpose annotations for this kind of thing in the past, but here I guess I would propose a new `@DurationStorage` annotation, with `enum TimeZoneStorageType { NANOSECOND, SECOND, DAY }`.

This would actually be pretty straightforward to implement: all the machinery is there. I was very pleasantly surprised to notice that actually implemented HQL duration support in a pretty robust way, where durations know about their own precision and the translation code automatically adapts to the precision of the duration. So that's easy enough.

Finally, I believe that @beikov would like to be able to store a `Duration` in a SQL `interval second` column, and I think we can accommodate that here too, though it would take a bit more work to get it working in HQL. (What was there previously was _not_ working.) Ultimately we could have something like:

    enum DurationStorageType { 
        NANOSECOND, 
        SECOND, 
        DAY,
        INTERVAL_SECOND
    }

Alternatives? 

- Well, one straightforward alternative worth due consideration is to just introduce `SecondDurationJavaType` and `DayDurationJavaType`, and use the `@JavaType` annotation. This would work well-enough, I believe, and I guess it could also be extended to accommodate `IntervalDurationJavaType`. Note, however, that the HQL translator would need to be aware of these new `JavaType`s and special-case them using `instanceof`, so it's not _quite_ as clean as it looks.
- Perhaps this can even be handled with just a converter? I would have to try. (Hell, it probably could even be done with a `JdbcType`, though I think that would be an abuse of the APIs.)

Either of these alternative approaches has the advantage of not introducing new annotations, though on the flip side that makes the functionality less-discoverable to users. And I guess we could argue about which of:

    @DurationStorage(DAY)
    Duration duration;

vs

    @JavaType(DayDurationJavaType.class)
    Duration duration;

is more aesthetically pleasing. I'm not sure, but I guess I lean toward the first option. Especially since we have already gone there with `@TimeZoneStorage`.

**Comment by Sanne** [comment]
That looks quite nice, and I have no objections at all in adding such handy mapping annotations - however let me point out that I don't agree with the premise of "user friendlyness" of not storing nanoseconds.

Encoding such values at high resolution (not precision) at least it ensures that on reading a stored value, the same value is returned.  Let's remember that time-zone dependent values are a mess as they depend on politics and systems being updated timely and consistently to be aware of whatever weird non-technical decisions politicians made; even assuming time-zone settings are correct from a technical point of view, it's quite possible that the JVM and the DB are at different stages or quality of patching in such regard (i.e. one might not have been updated to the latest oddness yet), and since storing at lower precision implies truncations based on such inconsistent rules, the "user friendliness" is arguable. Storing at highest resolution at least prevents some classes of such oddities.

Also, most users won't care what you store. So while I like being able to control the resolution I'd keep high resolution as a default.

**Reply by gavinking** [reply]
> however let me point out that I don't agree with the premise of "user friendlyness" of not storing nanoseconds.

I'm not suggesting not storing them. `numeric(21,9)` lets you store nanoseconds as the fractional part, after the decimal point. And since `numeric` is an exact type, there's no rounding error.

UPDATE: this isn't quite right, as @yrodiere points out below, for `DAY` storage, there can be a rounding error, it looks like `numeric` is actually a fair bit _worse_ than `double` in this respect.

On the other hand, `SECOND` storage _doesn't_ have rounding error.

**Reply by Sanne** [reply]
great!

**Comment by yrodiere** [comment]
> but it's not very nice to open up the database and see a value with 11 zeros on it representing one measly hour

I'd argue that's an endless problem: when I'm expressing a duration in years, the number of days will likely be so large it will feel meaningless.

> Or, hell, even as days would make sense, using something like numeric(22,14).

I very much doubt all durations expressed as a fraction of a day can be represented as a decimal number (with a finite number of decimal digits). You're bound to hit rational but non-decimal numbers at some point and then you might lose precision. That's the kind of limitation I'd hate to expose users to, especially just for the convenience of reading "understandable" number in the DB, and especially if the problem can be introduced by something so seemingly innocuous as picking one enum value over another.

For example take a duration representing a prime number of seconds, say `32212254719`. In days (divided by `86400`), that's `372827.0222106481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481481...`.

Rounding might work out fine in this case (I don't know), but then it might not in other cases.

> `enum TimeZoneStorageType { NANOSECOND, SECOND, DAY }.`

I suppose you meant `DurationStorageType`? I don't see what Timezone has to do with this.

> Alternatives?

I believe either alternative is preferable to `@DurationStorage`, be it only because the separation between "safe" storage and unsafe storage such as "days" is clearer.

**Reply by yrodiere** [reply]
Note if you introduce a concept of "resolution" instead of "storage type" (i.e. you just truncate the duration to the selected resolution: days, seconds, years, ...), the problem of rational, non-decimal numbers goes completely away, and at least it would be very clear that there can be data loss if you're not careful (and we could even throw an exception when that happens).

That might be a better solution, since I don't understand why someone would want to store a duration as days but still keep around a small additional number of seconds. They might as well store that duration as seconds in that case.

**Reply by gavinking** [reply]
> I'd argue that's an endless problem: when I'm expressing a duration in years, the number of days will likely be so large it will feel meaningless.

Remember that day-second durations are semantically different to year-month durations, and a Java `Duration` is fundamentally a day-second duration. It _cannot be used to represent an amount of years_.

But sure, going to your larger point: that's the exactly motivation for giving people a choice here. Depending on what a particular `Duration` represents, you should use different units to report it. 

I mean, for example, in physics you don't see all values quoted in eV. Sometimes we use MeV, sometimes GeV, sometimes TeV. Because it's really hard to read a number with nine-ten zeros on the end of it and _not makes a mistake_.

>I very much doubt all durations expressed as a fraction of a day can be represented as a decimal number (with a finite number of decimal digits)

OK, so a day is ~10^14 nanoseconds.

And we've actually been through this before: _there are no measurements with an accuracy of 13 decimal digits_, except for maybe [this one](https://en.wikipedia.org/wiki/Electron_magnetic_moment), the great outstanding triumph of modern physics.

If you ever see a number with more than 13 digits of precision, it's almost certainly lying to you. In business computing you do not need that amount of precision _ever_. (In fact, you essentially never even need it even in scientific computing.)

But sure, you shouldn't use `DAY` if you actually have a number that is precise to seconds or microseconds or nanoseconds. You would use `DAY` if you have a number that is precise to minutes at most. That's, like, pretty clear, I think.

> I suppose you meant `DurationStorageType`? 

Yes, sorry, copy/paste error.

**Reply by gavinking** [reply]
(Remember, @yrodiere, a quantity like "120 ns" is _always_ approximate value to begin with. It's never _exactly_ 120 ns, that's an impossibility.)

**Reply by gavinking** [reply]
Wait, perhaps I misunderstood your concern. Is your concern reversibility of the conversion to days and back, _when you have something like an exact quantity of hours_?

~I'm inclined to think that with that amount of overhead you would be perfectly fine. Like, if I take your example and try it out in postgres:~

```
hreact=# create table t(t numeric(22,14));
CREATE TABLE
hreact=# insert into t values(32212254719/86400);
INSERT 0 1
hreact=# select t*86400 from t;
          ?column?          
----------------------------
 32212252800.00000000000000
(1 row)
```
~Looks good.~


Same Java `double`s, FTR:

```
jshell> double x = 32212254719.0
x ==> 3.2212254719E10

jshell> double y=86400;
y ==> 86400.0

jshell> double z = 86400;
z ==> 86400.0

jshell> (x/y)*z
$4 ==> 3.2212254719E10
```

But OK, I guess that's something I should look into more closely. I might have got lucky in this particular case. It certainly does seem like in the case of FP I could get a number with `.9999999999999999` on the end of it, though I'm not sure, maybe the FP spec is set up to avoid that. And I'm definitely not certain what happens with exact decimals.

**Reply by gavinking** [reply]
> Looks good.

Ah hahah, wait, no it doesn't. What?? Let me try again.

**Reply by gavinking** [reply]
OK, fine, you're right @yrodiere, `numeric` on postgres is significantly worse at this than `double`. This I did not know.

So, fine, `DAY` storage probably isn't a great idea.

```
hreact=# create table t(t double precision);
CREATE TABLE
hreact=# insert into t values(32212254719.0/86400);
INSERT 0 1
hreact=# select t*86400 from t;
  ?column?   
-------------
 32212254719
(1 row)
```

Good.

```
hreact=# create table t(t numeric(24,15));
CREATE TABLE
hreact=# insert into t values(32212254719::numeric/86400::numeric);
INSERT 0 1
hreact=# select t*86400::numeric from t;
          ?column?           
-----------------------------
 32212254718.999999987200000
(1 row)
```

Bad, as you predicted.

**Reply by Sanne** [reply]
Reversability of the function being able to get back to the exact same representation would be my primary concern; and not because it could be technically wrong (I doubt that) but because of user surprise.

No biggie, I suppose we could warn about that possibility.  I just don't want to have to explain that to hundreds of users who don't have basic math training :)

**Reply by yrodiere** [reply]
> Wait, perhaps I misunderstood your concern. Is your concern reversibility of the conversion to days and back, when you have something like an exact quantity of hours?

Indeed, that was my concern. Users are not introducing the mistake here, we are.

> No biggie, I suppose we could warn about that possibility

I doubt documentation will cut it here, especially if we make it so easy to switch from `SECONDS` to `NANOSECONDS` to `DAYS`. It just feels surprising that one of those carries such limitations, especially if those limitations are not consistent (only for some prime numbers etc.).

Like I said, the alternative solutions might be easier to document effectively (read: with users actually getting warned before using the feature).

OR we could just *always* truncate values, so that at least the behavior is consistent. It feels "right" to me that users needing nanosecond precision will end up storing an (integer) number of nanoseconds while users needing only day/second precision will end up storing an (integer) number of days/seconds. And if we name the annotations/enums right, the truncation won't be surprising at all.

**Reply by gavinking** [reply]
So it looks like it works as long as you call `round()` explicitly. (Tried on both postgres and MySQL.)

```
hreact=# select round(t*86400) from t;
    round    
-------------
 32212254719
```

Which is fine. But the whole point of this was to get values displayed nicely to the user in adhoc queries, where they won't (by default) know to use `round()`. So that undermines the value of having this at least somewhat.

**Comment by beikov** [comment]
So it seems to me that we all agree the following options would be useful:

```
enum DurationStorageType { 
    NANOSECOND, // store as numeric(21)
    SECOND, // store as numeric(21,9)
    INTERVAL_SECOND // store as interval second
}
```

I think that a truncating `DAY` option could also make sense, similar to how `@Temporal(DATE)` would truncate the time part away, but maybe we should rather name it `DAY_TRUNCATE` if we want that and store it as `bigint`?

Apart from that, the idea to have a storage type enum kind of would help mapping `Period` to other interval types (also see https://github.com/hibernate/hibernate-orm/discussions/4892)
